File: pat1066.cpp
/*
	思路：就是一般的平衡二叉树建立过程，网上可以搜一下教程

*/

#include<iostream>
#include<algorithm>
using namespace std;

struct node{
	
	int value, depth = 1;
	node* left = NULL;
	node* right = NULL;
};

int getHight(node* root) {

	if (root == NULL) {
		return 0;
	}
	else {
		return root->depth;
	}
}

void updateHight(node* root) {
	root->depth = max(getHight(root->left),
		getHight(root->right)) + 1;
}

int getBalanceFactor(node* root) {
	return getHight(root->left) - getHight(root->right);
}

void leftRotate(node* &root) {
	node* temp = root->right;
	root->right = temp->left;
	temp->left = root;
	updateHight(root);
	updateHight(temp);
	root = temp;
}

void rightRotate(node* &root) {
	node* temp = root->left;
	root->left = temp->right;
	temp->right = root;
	updateHight(root);
	updateHight(temp);
	root = temp;
}

void insert(node* &root, int value) {
	
	if (root == NULL) {
		root = new node();
		root->value = value;
	}
	else if (value < root->value) {
		insert(root->left, value);
		updateHight(root);
		if (getBalanceFactor(root) == 2) {
			if (getBalanceFactor(root->left) == 1) {
				rightRotate(root);
			}
			else {
				leftRotate(root->left);
				rightRotate(root);
			}

		}
	}
	else {
		insert(root->right, value);
		updateHight(root);
		if (getBalanceFactor(root) == -2) {
			if (getBalanceFactor(root->right) == -1) {
				leftRotate(root);
			}
			else {
				rightRotate(root->right);
				leftRotate(root);
			}
		}
	}
}


int main() {

	int N;
	scanf("%d", &N);
	node* root = NULL;
	for (int i = 0; i < N; i++) {
		int k;
		scanf("%d", &k);
		insert(root, k);
	}

	printf("%d", root->value);
	return 0;
}

File: pat1066.py
"""
    思路：就是一般的平衡二叉树建立过程，网上可以搜一下教程

"""



class node():

    def __init__(self,value):
        self.value = value
        self.left = None
        self.right = None
        self.depth = 1
    
def getHight(root):
    if root is None:
        return 0;
    else:
        return root.depth

def updateHight(root):
    root.depth = max(getHight(root.left), getHight(root.right)) + 1

def getBalanceFactor(root):
    return getHight(root.left) - getHight(root.right)


def leftRotate(root):
    temp = root.right
    root.right  =temp.left
    temp.left = root
    updateHight(root)
    updateHight(temp)
    root = temp
    return root

def rightRotate(root):
    temp = root.left
    root.left = temp.right
    temp.right = root
    updateHight(root)
    updateHight(temp)
    root = temp
    return root

def insert(root, value):
    if root is None:
        root = node(value)

    elif value < root.value:
        root.left = insert(root.left, value)
        updateHight(root)
        if getBalanceFactor(root) == 2:
            if getBalanceFactor(root.left) == 1:
                root = rightRotate(root)
            else:
                root.left = leftRotate(root.left)
                root = rightRotate(root)
    elif value > root.value:
        root.right = insert(root.right, value)
        updateHight(root)
        if getBalanceFactor(root) == -2:
            if getBalanceFactor(root.right) == -1:
                root = leftRotate(root)
            else:
                root.right = rightRotate(root.right)
                root = leftRotate(root)

    return root



N = int(input())
origin = [int(x) for x in input().split()]

root = None
for value in origin:
    root = insert(root, value)

print(root.value)